index.bs

```diff
@@ -10518,126 +10522,140 @@ The characteristics of this property are as follows:
-                this property corresponding to the attribute appears on.
-            1.  Let |O| be the <emu-val>this</emu-val> value.
-            1.  If |O| is a [=platform object=],
-                then [=perform a security check=], passing:
-                *   the platform object |O|,
-                *   the [=identifier=] of the [=attribute=], and
-                *   the type “setter”.
-            1.  Let |validThis| be <emu-val>true</emu-val> if |O| is a
-                [=platform object=] that implements |I|, or
-                <emu-val>false</emu-val> otherwise.
-            1.  If |validThis| is <emu-val>false</emu-val> and the
-                [=attribute=] was not specified with the
-                [{{LenientThis}}] [=extended attribute=],
-                then [=ECMAScript/throw=] a <emu-val>TypeError</emu-val>.
-            1.  If the attribute is declared with a [{{Replaceable}}]
-                extended attribute, then:
-                1.  Let |P| be the identifier of the attribute.
-                1.  Call [=CreateDataProperty=](|O|, |P|, |V|).
+        1.  If |attribute|'s type is a [=promise type=], then:
+            1.  Let |reject| be the initial value of [=%Promise%=].reject.
+            1.  Return the result of calling |reject| with [=%Promise%=] as the
+                <emu-val>this</emu-val> object and the exception as the single argument value.
+        1.  Otherwise, end these steps and allow the exception to propagate.
+    1.  Let |F| be [=!=] [=CreateBuiltinFunction=](|realm|,
+        |steps|, the [=%FunctionPrototype%=] of |realm|).
+    1.  Let |name| be the string "get " prepended to |attribute|'s [=identifier=].
+    1.  Perform [=!=] [=SetFunctionName=](|F|, |name|).
+    1.  Perform [=!=] [=DefinePropertyOrThrow=](|F|, "length",
+        PropertyDescriptor<span class="prop-desc">{\[[Value]]: 0, \[[Writable]]: <emu-val>false</emu-val>, \[[Enumerable]]: <emu-val>false</emu-val>, \[[Configurable]]: <emu-val>true</emu-val>}</span>).
+    1. Return |F|.
+
+</div>
+
+<div algorithm>
+
+    The <dfn id="dfn-attribute-setter" export>attribute setter</dfn> is created as follows, given an
+    [=attribute=] |attribute|, an [=interface=] |target|, and a [=Realm=] |realm|:
+
+    1.  If |attribute| is [=read only=] and does not have a
+        [{{LenientThis}}], [{{PutForwards}}] or [{{Replaceable}}] [=extended attribute=], return
+        <emu-val>undefined</emu-val>; there is no [=attribute setter=] function.
+    1.  Assert: |attribute|'s type is not a [=promise type=].
+    1.  Let |steps| be the following series of steps:
+        1.  If no arguments were passed, then
+        1.  Let |V| be the value of the first argument passed.
+        1.  Let |id| be |attribute|'s [=identifier=].
+        1.  Let |O| be <emu-val>null</emu-val>.
+        1.  If |attribute| is not a [=static attribute=]:
+            1.  If the <emu-val>this</emu-val> value is <emu-val>null</emu-val> or
```

Comment:
If find all of the substeps of this step 5 super difficult to parse. I don't doubt that the algorithm is correct, but I really struggle to understand its intent. I can't imagine modifying something in here without accidentally breaking stuff.

Here's a proposed alternative (which is slightly less DRY, but hopefully a bit clearer). I added an optional note (prefixed with ?) to clarify the exit point of the substep. Maybe that helps (or is over the top—I'm not sure).

```
1.  If |attribute| is a [=regular attribute=]:
    1.  If the <emu-val>this</emu-val> value is <emu-val>null</emu-val> or
        <emu-val>undefined</emu-val>, set |O| to |realm|'s [=Realm/global object=].
        (This will subsequently cause a <emu-val>TypeError</emu-val> in a few steps, if
        the global object does not implement |target| and [{{LenientThis}}] is not
        specified.)
        <!-- https://www.w3.org/Bugs/Public/show_bug.cgi?id=18547#c9 -->
    1.  Otherwise, set |O| to the <emu-val>this</emu-val> value.
    1.  If |O| is a [=platform object=], then [=perform a security check=], passing |O|,
        |id|, and "setter".
     1.  Let |validThis| be `true` if |O| is a [=platform object=] that
         implements the interface |target|, or `false` otherwise.
     1.  If |validThis| is `true`, then:
         1.  If |attribute| is declared with a [{{LenientSetter}}] extended attribute, then:
             1.  Return <emu-val>undefined</emu-val>.
         1.  If |attribute| is declared with a [{{PutForwards}}] extended attribute, then:
             1.  Let |Q| be [=?=] [=Get=](|O|, |id|).
             1.  If [=Type=](|Q|) is not Object, then [=ECMAScript/throw=] a
                 <emu-val>TypeError</emu-val>.
             1.  Let |forwardId| be the identifier argument of the [{{PutForwards}}] extended
                 attribute.
             1.  Perform [=?=] [=Set=](|Q|, |forwardId|, |V|).
             1.  Return <emu-val>undefined</emu-val>.
         1.  If |attribute| is declared with the [{{Replaceable}}] extended attribute, then:
             1.  Perform [=?=] [=CreateDataProperty=](|O|, |id|, |V|).
             1.  Return <emu-val>undefined</emu-val>.
        
        ? Note: setters who are not special-cased here will
        ? continue being defined in the parent steps
        ? along with static attributes.
     1.  Otherwise if |attribute| was specified with the [{{LenientThis}}] [=extended attribute=],
         1.  If |attribute| is declared with the [{{Replaceable}}] extended attribute,
             then perform [=?=] [=CreateDataProperty=](|O|, |id|, |V|).
         1. Return <emu-val>undefined</emu-val>.
     1.  Otherwise, [=ECMAScript/throw=] a <emu-val>TypeError</emu-val>.
```